src/utils/car-block-getter.js

```diff
@@ -0,0 +1,51 @@
+import { CarBlockIterator } from '@ipld/car/iterator'
+import toIterable from 'browser-readablestream-to-it'
+
+import { verifyBlock } from './car.js'
+import { promiseTimeout } from './timers.js'
+import { TimeoutError, VerificationError } from './errors.js'
+
+// Assumptions
+// * client and server are both using DFS traversal.
+// * Server is sending CARs with duplicate blocks.
+export class CarBlockGetter {
+  constructor (carItr, opts = {}) {
+    this.carItr = carItr
+    this.getBlockTimeout = opts.getBlockTimeout ?? 1_000 * 10
+  }
+
+  static async fromStream (carStream) {
+    const iterable = await CarBlockIterator.fromIterable(
+      asAsyncIterable(carStream)
+    )
+    const carItr = iterable[Symbol.asyncIterator]()
+    return new CarBlockGetter(carItr)
+  }
+
+  async get (cid, options) {
+    const { value, done } = await promiseTimeout(
+      this.carItr.next(),
+      this.getBlockTimeout,
+      new TimeoutError(`get block ${cid} timed out`)
+    )
```

Comment:
This might be handled higher up the stack (e.g. in `ipfs-unixfs-exporter`), but it'd probably be a good idea to explicitly mentioned identity CIDs. In particular,
1. Add a test fixture that verifies you can handle traversing a DAG that has some (but isn't entirely) CIDs with identity multihashes
2. If this is expected to be handled by consumers of this class then it should be documented on the class

Reply:
How does this break on identity CIDs? Will CAR files typically not contain blocks for identity CIDs?

Reply:
Is there a repo of CAR files that facilitate testing different scenarios? I find it tedious to create CAR files with specific quirks.

Reply:
We've tried to rule them out from trustless gateway https://specs.ipfs.tech/http-gateways/trustless-gateway/#car-dups-content-type-parameter, they're a scourge on the CAR format so if we can exclude them then that's good.

The catch is that whenever you implement a blockstore-like interface, you probably should handle them anyway because they can appear in the data, even if not as extra explicit blocks in the CAR, and some layer needs to be able to handle them properly. The best place is usually at the blockstore since it's the choke point of blocks; a `GET` operation should just look for `0x0` and say "yep, here's the block you asked for" and give the block right back extracted out of the CID. It's annoying, but it's going to come up at some point and catch you by surprise ... this is one of the inevitabilities of identity CIDs sadly.

Reply:
added a check for identity cids. it looks like `ipfs-unixfs-exporter` handles them as well.